In [ ]:
!pip install openai -q
!!pip3 install -qU \
    langchain \
    pinecone-client

!pip3 install protobuf==3.20.3 -q
!pip3 install apache-beam==2.50.0 -q
!!pip install gradio -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 221.4/221.4 kB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.0/75.0 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.9/76.9 kB 6.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 4.1 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.
llmx 0.0.15a0 requires tiktoken, which is not installed.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.7/14.7 MB 53.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.7/89.7 kB 10.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.7/138.7 kB 15.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 152.0/152.0 kB 16.9 MB/s eta 0:00:00
  Preparing metadata (

['\x1b[?25l     \x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b \x1b0.0/16.6 MB\x1b \x1b?\x1b eta \x1b-:--:--\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b \x1b0.1/16.6 MB\x1b \x1b2.6 MB/s\x1b eta \x1b0:00:07\x1b',
 '\x1b[2K     \x1b━\x1b\x1b╸\x1b\x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b \x1b0.7/16.6 MB\x1b \x1b10.9 MB/s\x1b eta \x1b0:00:02\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━\x1b\x1b╺\x1b\x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b \x1b4.2/16.6 MB\x1b \x1b40.6 MB/s\x1b eta \x1b0:00:01\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━━━━━━━━━━━━━━━\x1b\x1b╸\x1b\x1b━━━━━━━━━━━━━━━\x1b \x1b10.3/16.6 MB\x1b \x1b91.2 MB/s\x1b eta \x1b0:00:01\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b\x1b╸\x1b \x1b16.5/16.6 MB\x1b \x1b190.4 MB/s\x1b eta \x1b0:00:01\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b\x1b╸\x1b \x1b16.5/16.6 MB\x1b \x1b190.4 MB/s\x1b eta \x1b0:00:01\x1b',
 '\x1b[2K     \x1b━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b \x1b16.6/16.6 

In [ ]:
!pip install tiktoken -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 19.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.


In [ ]:
import gradio as gr

In [ ]:
from google.colab import userdata
open_ai = userdata.get('OPENAI_API_KEY')
pinecone_api = userdata.get('PINECONE_API')

In [ ]:
import os
import openai
from openai import OpenAI
import tiktoken
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.text_splitter import RecursiveCharacterTextSplitter
import pandas as pd
import pinecone
from langchain.chains.router import MultiPromptChain
from langchain.chains.router.llm_router import LLMRouterChain,RouterOutputParser
from langchain.prompts import ChatPromptTemplate
from langchain.chat_models import ChatOpenAI
from langchain.chains import ConversationChain
from langchain.memory import ConversationBufferMemory
from langchain.memory import ConversationBufferWindowMemory
from langchain.chains import LLMChain
from langchain.prompts import PromptTemplate

In [ ]:
os.environ['OPENAI_API_KEY'] = open_ai  #Initialize API Key
client = OpenAI()

In [ ]:

YOUR_API_KEY = pinecone_api
# Find ENV (cloud region) next to API key in console
YOUR_ENV = 'gcp-starter'

INDEX_NAME = 'temp'
# Initialize Pinecone client
pinecone.init(
    api_key=YOUR_API_KEY,
    environment=YOUR_ENV
)
demo_index = pinecone.Index('temp')

In [ ]:
model_name = 'text-embedding-ada-002'
embed = OpenAIEmbeddings(
    model=model_name,
)

# Agent Convert **ROMAN URDU** into **English**

In [ ]:
 def urdu_to_English(messages):
      response = client.chat.completions.create(
                model="gpt-3.5-turbo",
                        messages=messages,
                        temperature=1,
                        max_tokens=256,
                        top_p=1,
                        frequency_penalty=0,
                        presence_penalty=0
                )
      return response.choices[0].message.content



In [ ]:
# Prompt Template for translation.
messages = [
            {
              "role": "system",
              "content": "You are a  exceptional Roman Urdu to English Translator.  You translate roman Urdu sentences into English sentences."
            },
            {
              "role": "user"

            }
        ]

In [ ]:
sentence  = "meri age 28 saal ha or me aik aurat hu Mery sir me drd ha or 2 din sy bukhar ha or weakness feel ho rhi ha. Kl dawai ly thi pr uska koi faida nai howa. \
Is sy pehly  aesy nai hota tha."
messages[1]['content'] = sentence
messages

[{'role': 'system',
  'content': 'You are a  exceptional Roman Urdu to English Translator.  You translate roman Urdu sentences into English sentences.'},
 {'role': 'user',
  'content': 'meri age 28 saal ha or me aik aurat hu Mery sir me drd ha or 2 din sy bukhar ha or weakness feel ho rhi ha. Kl dawai ly thi pr uska koi faida nai howa. Is sy pehly  aesy nai hota tha.'}]

In [ ]:
pinecone_query = urdu_to_English(messages)
pinecone_query

"I am 28 years old and I am a woman. I have a headache and have had a fever for the past 2 days. I am also feeling weak. I took medicine yesterday but it didn't help. This hasn't happened to me before."

# **Data Generator Agent**

In [ ]:
prompt = "Your a medical data generator.\
         You  will generate medical data of given disease.\
          Data contain Patient's History, Patients are  female.Records  of body vitals like weight,\
           blood pressure and other body vital data. And description of patient. Data must be in JSON format. \
           JSON keys are History, Body vitals and Descripton of Patient. And values must be in the form of Paragraph ."

data = [
            {
              "role": "system",
              "content": prompt
            },
            {
              "role": "user"

            }
        ]

In [ ]:
data[1]['content'] = "Disease name is 'Dengue Fever'"
data

[{'role': 'system',
  'content': "Your a medical data generator.         You  will generate medical data of given disease.          Data contain Patient's History, Patients are  female.Records  of body vitals like weight,           blood pressure and other body vital data. And description of patient. Data must be in JSON format.            JSON keys are History, Body vitals and Descripton of Patient. And values must be in the form of Paragraph ."},
 {'role': 'user', 'content': "Disease name is 'Dengue Fever'"}]

In [ ]:
for i in range(3):
    data_response = client.chat.completions.create(
                    model="gpt-3.5-turbo",
                            messages=data,
                            temperature=1,
                            max_tokens=1024,
                            top_p=1,
                            frequency_penalty=0,
                            presence_penalty=0
                    )
    respr = data_response.choices[0].message.content
    print(respr.split("\n"))

['{', '  "History": "The patient, a 35-year-old female, has a history of traveling to a dengue endemic area in the past week. She started experiencing symptoms such as high fever, severe headache, muscle and joint pain, nausea, vomiting, and rash. The patient sought medical attention promptly, and laboratory tests confirmed the diagnosis of dengue fever.",', '', '  "Body Vitals": "Over the course of the illness, the patient\'s weight remained stable at 65 kilograms. Her body temperature fluctuated between 38.5 to 40 degrees Celsius, indicating a persistent high-grade fever. The patient\'s blood pressure was slightly lower than usual, with readings around 100/70 mmHg. Additionally, there was evidence of thrombocytopenia, with a platelet count dropping below 100,000 cells/mm³.",', '', '  "Description": "The patient appears pale and fatigued, with sunken eyes and dry lips. She complains of generalized body aches and fatigue, which worsen with movement. The rash, known as the \'dengue rash

['{',
 '  "History": "The patient, a 35-year-old male, visited the clinic with complaints of high fever, severe headache, and body aches. He mentioned that the symptoms started around a week ago. The patient has a history of travelling to an area known for dengue fever outbreaks. He reported experiencing episodes of nausea and vomiting, along with a loss of appetite. There was no history of any major illness in the past.",',
 '  "Body Vitals": "Upon examination, the patient\'s body temperature was found to be 102°F. His blood pressure was measured at 120/80 mmHg. The patient\'s weight was recorded as 70 kilograms. Other body vital observations included a rapid heartbeat and a slightly lowered platelet count.",',
 '  "Description": "The patient appears weak and fatigued. He has pale skin and complains of dizziness when standing up. His spleen and liver are not enlarged. The patient\'s blood test confirmed the presence of dengue fever antibodies. He has been advised to take plenty of flu

'{',
 '  "History": "The patient has a history of recent travel to a tropical or subtropical region where dengue fever is endemic. They may have been exposed to the dengue virus through mosquito bites during their travel. Dengue fever is a viral infection transmitted by the Aedes mosquito.",',
 '  "Body vitals": "The patient\'s body temperature may rise to around 104°F (40°C) during the initial phase of dengue fever. Additionally, they may experience intense headache, joint and muscle pain, rash, and pain behind the eyes. As the fever progresses, their blood platelet count may decrease, leading to an increased risk of bleeding.",',
 '  "Description of Patient": "The patient is presenting symptoms of dengue fever, such as high fever, severe headache, and body pain. They may have a rash on their body and complain of pain behind the eyes. The patient\'s blood platelet count is being monitored regularly to assess their condition and ensure early detection of any complications. Adequate hydration and rest are being provided to support the patient\'s recovery."',
 '}'

 '{',
 '  "History": "The patient, a 35-year-old male, visited the clinic with complaints of high fever, severe headache, and body aches. He mentioned that the symptoms started around a week ago. The patient has a history of travelling to an area known for dengue fever outbreaks. He reported experiencing episodes of nausea and vomiting, along with a loss of appetite. There was no history of any major illness in the past.",',
 '  "Body Vitals": "Upon examination, the patient\'s body temperature was found to be 102°F. His blood pressure was measured at 120/80 mmHg. The patient\'s weight was recorded as 70 kilograms. Other body vital observations included a rapid heartbeat and a slightly lowered platelet count.",',
 '  "Description": "The patient appears weak and fatigued. He has pale skin and complains of dizziness when standing up. His spleen and liver are not enlarged. The patient\'s blood test confirmed the presence of dengue fever antibodies. He has been advised to take plenty of fluids, rest, and monitor his platelet count regularly to ensure early detection of any complications."',
 '}'


['{', '  "History": "Patient has a history of recent travel to a tropical region where dengue fever is prevalent. The patient may have been exposed to the dengue virus through mosquito bites during their travels. It is important to note that dengue fever is not contagious and cannot be spread from person to person.",', '  "Body Vitals": "Weight: 70 kg\\nBody temperature: 38.5°C\\nBlood pressure: 120/80 mmHg\\nHeart rate: 90 bpm\\nRespiratory rate: 16 breaths per minute\\nBlood sugar level: 110 mg/dL",', '  "Description of Patient": "The patient is a 35-year-old male who presented with symptoms such as high fever, severe headache, joint and muscle pain, rash, and fatigue. These symptoms are characteristic of dengue fever. The patient reports feeling weak and experiencing an overall sense of malaise. They have no known underlying medical conditions and have not taken any specific medications prior to the onset of symptoms."', '}']

['{', '  "History": "The patient is a 38-year-old female who presented with symptoms of dengue fever. She reported experiencing sudden onset of high fever, severe headache, joint and muscle pain, and rash. The patient recently returned from a tropical vacation, where she may have been exposed to mosquito bites. She has no significant medical history or known allergies.",', '  ', '  "Body Vitals": "The patient\'s weight is 65 kg. Her blood pressure is within normal range with a systolic pressure of 120 mmHg and a diastolic pressure of 80 mmHg. She has a slightly elevated heart rate of 90 beats per minute. The patient\'s body temperature is 38.5 degrees Celsius. Blood tests revealed a decreased platelet count and an increased hematocrit level, indicating a possible dengue infection.",', '  ', '  "Description": "Upon physical examination, the patient appears fatigued and weak. She has a petechial rash on her arms and legs, which is characteristic of dengue fever. The patient complains of severe joint and muscle pain, particularly in her knees and elbows. She also experiences nausea, vomiting, and loss of appetite. The doctor suspects a dengue infection based on the patient\'s symptoms and travel history, and further tests will be conducted to confirm the diagnosis."', '}']


['{', '  "History": "The patient with Dengue Fever presented with a history of sudden onset high-grade fever, severe headache, joint and muscle pain, and rash. The fever typically lasts for 5-7 days and is accompanied by other symptoms such as nausea, vomiting, and abdominal pain. The patient also reported a history of mosquito bite exposure in an endemic area.",', '  "Body Vitals": "The patient\'s body vitals showed an elevated body temperature of 38.5°C (101.3°F), increased heart rate of 100 beats per minute, and a slightly lowered blood pressure of 110/70 mmHg. Laboratory analysis revealed a reduced platelet count of 100,000 cells/mm³ and an increased hematocrit level of 45%. The patient\'s weight was recorded as 55 kg.",', '  "Description of Patient": "The patient is a 28-year-old female who has recently returned from a tropical vacation. She describes her headache as severe and throbbing, and her joint and muscle pain as debilitating. The rash appeared as small red spots on her arms and legs. The patient is experiencing fatigue and overall weakness. She has been advised to take plenty of fluids and rest, and to monitor her body temperature regularly."', '}']


['{', '  "History": "The patient\'s history indicates that they recently visited a tropical region where dengue fever is prevalent. They may have been exposed to the dengue virus through mosquito bites. It is important to note that dengue fever is not a contagious disease and cannot be spread from person to person.",', '  "Body vitals": "The patient\'s body vitals include a body temperature of 102°F, elevated heart rate of 110 beats per minute, and a drop in blood pressure to 90/60 mmHg. These changes in body vitals are common in cases of dengue fever. The patient may also experience symptoms such as severe headache, joint and muscle pain, fatigue, and a rash.",', '  "Description of Patient": "The patient is a 35-year-old male who presented with high fever, body aches, and a rash. They have a history of travel to a dengue-endemic area in the past week. Upon physical examination, the patient appears pale and fatigued. They complain of severe headache, particularly behind the eyes, and have reported joint and muscle pain. The presence of a rash is observed on the patient\'s arms and legs. Laboratory tests confirmed the diagnosis of dengue fever']


['{', '  "Disease": "Dengue Fever",', '  "Patient History": "The patient, a 32-year-old male, presented with a sudden onset of high fever, severe headache, and body aches. He reported having traveled to a tropical region two weeks ago. The fever was accompanied by intense joint pain, especially in the wrists and ankles. The patient also experienced fatigue and mild nausea. He reported no known allergies or prior history of dengue fever.",', '  "Body Vitals": "Upon examination, the patient\'s temperature was recorded as 39.2°C (102.5°F). His blood pressure was slightly low at 100/70 mmHg. The patient appeared pale and dehydrated. Laboratory tests revealed a decreased platelet count of 90,000 cells/µL (normal range: 150,000-450,000 cells/µL) and a hematocrit level of 42% (normal range: 38-54%). The patient\'s white blood cell count was within the normal range. Additionally, the patient\'s liver enzymes were elevated, indicating liver involvement.",', '  "Description": "Dengue fever is a mosquito-borne viral infection that commonly occurs in tropical and subtropical regionns.']


['{', '  "Disease": "Dengue Fever",', '  ', '  "History": "The patient, a 42-year-old male, presented with symptoms of dengue fever, including high-grade fever, severe headache, joint and muscle pain, and a rash on the body. The patient mentioned that he had recently traveled to a tropical region where dengue fever is prevalent. He reported feeling fatigued and experiencing mild nausea. There was no history of dengue fever in the family.",', '', '  "Body Vitals": "During examination, the patient\'s weight was measured at 75 kg. His blood pressure was 120/80 mmHg. The patient had a pulse rate of 85 beats per minute and a temperature of 39.2°C. The doctor observed that the patient had petechiae on the lower limbs, which are small red or purple spots caused by bleeding under the skin. The platelet count was found to be significantly low, indicating thrombocytopenia, a common feature of dengue fever. The patient\'s liver enzymes were elevated and he had a mild drop in white blood cell count.",', '', '  "Description": "The patient exhibited classic symptoms and physical findings consistent with dengue fever. Dengue']


['{', '  "History": "The patient, a 32-year-old male, presents with a history of experiencing high fever, severe headache, joint and muscle pain, and rash. He reported that these symptoms started approximately five days ago after being bitten by a mosquito. The patient also mentioned feeling fatigued and having a loss of appetite. He visited a healthcare facility where blood tests confirmed a diagnosis of Dengue Fever.",', '  "Body Vitals": {', '    "Weight": "The patient\'s weight is recorded at 75 kilograms.",', '    "Blood Pressure": "The patient\'s blood pressure is within a normal range, with a reading of 120/80 mmHg.",', '    "Temperature": "The patient\'s body temperature is elevated, measuring at 39 degrees Celsius."', '  },', '  "Description": "The patient presents with classic symptoms of Dengue Fever, which is a mosquito-borne viral infection. Dengue Fever is characterized by high fever, severe headache, joint and muscle pain, and rash. It is commonly transmitted by the Aedes mosquito. The patient is experiencing fatigue and loss of appetite, which are typical manifestations of this disease. Timely diagnosis and appropriate management are crucial for the patient\'s recovery."', '}']


In [ ]:
history = []
body_vitals = []
description = []

for i in range(16):
  hist = input("Enter History")
  history.append(hist)
  body = input("Enter body vitals")
  body_vitals.append(body)
  desc = input("Enter Desc")
  description.append(desc)

Enter HistoryThe patient has a history of recent travel to a tropical or subtropical region where dengue fever is endemic. They may have been exposed to the dengue virus through mosquito bites during their travel. Dengue fever is a viral infection transmitted by the Aedes mosquito.
Enter body vitalsThe patient's body temperature may rise to around 104°F (40°C) during the initial phase of dengue fever. Additionally, they may experience intense headache, joint and muscle pain, rash, and pain behind the eyes. As the fever progresses, their blood platelet count may decrease, leading to an increased risk of bleeding.
Enter DescThe patient is presenting symptoms of dengue fever, such as high fever, severe headache, and body pain. They may have a rash on their body and complain of pain behind the eyes. The patient's blood platelet count is being monitored regularly to assess their condition and ensure early detection of any complications. Adequate hydration and rest are being provided to supp





['{', '  "History": "The patient, a 32-year-old male, has a history of travel to a dengue-endemic region. He visited a tropical country two weeks ago and started developing symptoms a few days after returning. He experienced sudden-onset high fever, severe headache, pain behind the eyes, joint and muscle pain, and a distinct rash. The patient sought medical attention and was diagnosed with dengue fever based on clinical symptoms and positive laboratory tests for dengue virus infection. He was advised to rest, stay hydrated, and monitor his symptoms closely.",', '', '  "Body Vitals": "During the initial assessment, the patient\'s weight was recorded to be 75 kilograms. His blood pressure was 120/80 mmHg, which was within the normal range. Other vital signs, such as heart rate and respiratory rate, were also within normal limits. However, his body temperature was elevated, reaching 39.5°C. The patient was closely monitored for any signs of worsening fever or other complications.",', '', '  "Description of Patient": "The patient appears weak and fatigued, complaining of persistent fever and body aches. He has a notable rash on his torso and extremities, consisting of small red spots. The patient reports']


['{', '  "History": "The patient, a 30-year-old male, presented with symptoms of dengue fever. He reported a recent history of travel to a dengue-endemic region. Two weeks prior to the onset of symptoms, he visited a tropical country where dengue fever is prevalent. He mentioned being bitten by mosquitoes frequently during his stay. The patient also stated that he had not received any dengue vaccine or taken any preventive measures against mosquito bites. He has no previous history of dengue fever or any other mosquito-borne diseases.",', '', '  "Body vitals": "The patient\'s weight upon admission was recorded as 75 kg. His blood pressure was measured to be 130/85 mmHg. Other body vital data includes a body temperature of 38.5°C, a pulse rate of 90 beats per minute, and a respiratory rate of 18 breaths per minute. Laboratory results showed a decrease in platelet count to 120,000 cells/mm³ and an increase in hematocrit level to 45%.",', '', '  "Description of Patient": "The patient presented with typical symptoms of dengue fever, including high-grade fever, severe headache, retro-orbital pain, joint and muscle a']


['{', '  "History": "The patient, a 45-year-old male, presented with a history of high fever for the past three days. Initially, he experienced sudden onset of severe headache, joint pain, and muscle pain. He also complained of fatigue and lack of appetite. The patient had no previous history of dengue fever or any other mosquito-borne diseases. He reported no recent travel history to dengue-endemic regions.",', '  ', '  "Body Vitals": "Upon admission, the patient\'s body vitals were assessed. He had a temperature of 39.5°C (103.1°F), which was accompanied by sweating and chills. His blood pressure was 120/80 mmHg, heart rate was 100 beats per minute, and respiratory rate was normal. The patient weighed 70 kg. Laboratory tests revealed a decreased platelet count of 80,000 cells per microliter, indicating thrombocytopenia, a common finding in dengue fever.",', '  ', '  "Description of Patient": "The patient appeared lethargic and was in mild distress due to severe headache and muscle pain. He had no visible rash or bleeding manifestations at the time of presentation. On physical examination, there was tenderness in the joints']

{', '  "History": "The patient, a 25-year-old female, presented with a history of fever, headache, and joint/muscle pain for the past week. She mentioned that she had recently returned from a tropical region where cases of dengue fever were reported. The patient reported feeling fatigued and experienced some mild dizziness. There was no significant past medical history of note.",', '  "Body Vitals": {', '    "Weight": "The patient\'s weight was recorded as 55 kg.",', '    "Blood Pressure": "The patient\'s blood pressure was measured at 120/80 mmHg.",', '    "Pulse Rate": "The patient\'s pulse rate was noted to be 80 beats per minute.",', '    "Body Temperature": "The patient\'s body temperature was found to be 38.5°C, indicating a fever."', '  },', '  "Description": "On physical examination, the patient appeared pale and fatigued. She had mild tenderness in the joints and muscle groups, especially in the lower back and legs. No rashes were noted on the skin. Laboratory tests revealed a low platelet count and elevated liver enzymes, which are characteristic findings in dengue fever. The patient was diagnosed with dengue fever and provided']


['{', '  "History": "The patient, a 35-year-old female, was admitted to the hospital with symptoms of dengue fever. She reported a sudden onset of high fever, severe headache, and joint and muscle pain. The patient mentioned that she had recently traveled to a tropical region where dengue is prevalent. She also complained of fatigue, loss of appetite, and nausea. The patient had no previous history of dengue fever.",', '  "Body Vitals": "Upon examination, the patient\'s temperature was recorded as 39.5°C (103.1°F), indicating a high-grade fever. Her blood pressure was slightly low, with a reading of 100/60 mmHg. The patient\'s heart rate was elevated at 100 beats per minute, and her respiratory rate was within the normal range at 16 breaths per minute. The doctor noticed petechial rashes on the patient\'s legs, which are characteristic of dengue fever.",', '  "Description": "The patient exhibited classical symptoms of dengue fever, including high fever, severe headache, joint and muscle pain, fatigue, and loss of appetite. Dengue fever is a viral infection transmitted by the Aedes mosquito. The incubation period for dengue fever']


['{', '  "History": "The patient, a 28-year-old female, has a history of dengue fever. She contracted the disease during a recent trip to a tropical region known for dengue outbreaks. She experienced symptoms such as high fever, severe headache, joint and muscle pain, rash, and fatigue. The dengue virus is transmitted by Aedes mosquitoes, which are common in tropical and subtropical areas. The patient was diagnosed with dengue fever based on her symptoms and a positive test for dengue antigen. She received medical treatment and recovered from the acute phase of the illness. However, she was advised to take precautionary measures to prevent future dengue infections, such as using insect repellents and eliminating mosquito breeding sites around her living area."', '', '  "Body Vitals": "During the acute phase of dengue fever, the patient experienced fluctuating body temperatures, with a high fever ranging from 101°F to 104°F. Her blood pressure also showed some variations, but remained within the normal range. The patient\'s weight remained stable throughout the illness, with no significant changes observed. Other body vital parameters, such as heart rate and respiratory rate, were within normal limits. The patient was monitored closely to ensure proper hydration and']

 '  "History": "The patient, a 35-year-old female, has a history of traveling to a dengue endemic area in the past week. She started experiencing symptoms such as high fever, severe headache, muscle and joint pain, nausea, vomiting, and rash. The patient sought medical attention promptly, and laboratory tests confirmed the diagnosis of dengue fever.",', '', '  "Body Vitals": "Over the course of the illness, the patient\'s weight remained stable at 65 kilograms. Her body temperature fluctuated between 38.5 to 40 degrees Celsius, indicating a persistent high-grade fever. The patient\'s blood pressure was slightly lower than usual, with readings around 100/70 mmHg. Additionally, there was evidence of thrombocytopenia, with a platelet count dropping below 100,000 cells/mm³.",', '', '  "Description": "The patient appears pale and fatigued, with sunken eyes and dry lips. She complains of generalized body aches and fatigue, which worsen with movement. The rash, known as the \'dengue rash,\' is present on the extremities and trunk, consisting of small red spots (petechiae) and maculopapular lesions. The patient reports feeling nauseous and has had episodes of vomiting. She also experiences a severe, throbbing headache, especially behind the eyes. Overall, the patient\'s condition is consistent with dengue fever, and appropriate measures are being taken for symptom management and fluid balance."', '}']




In [ ]:
dengue_df = {}
dengue_df['Description'] = description
dengue_df['History'] = history
dengue_df['Body_Vitals'] = body_vitals


In [ ]:
dengue_df = pd.DataFrame(dengue_df)

In [ ]:
dengue_df.head()

Description  \
0  The patient is presenting symptoms of dengue f...   
1  The patient appears weak and fatigued. He has ...   
2  The patient is a 28-year-old female who has re...   
3  Upon physical examination, the patient appears...   
4  The patient is a 35-year-old male who presente...   

                                             History  \
0  The patient has a history of recent travel to ...   
1  The patient, a 35-year-old male, visited the c...   
2  The patient with Dengue Fever presented with a...   
3  The patient is a 38-year-old female who presen...   
4  Patient has a history of recent travel to a tr...   

                                         Body_Vitals  
0  The patient's body temperature may rise to aro...  
1  Upon examination, the patient's body temperatu...  
2  The patient's body vitals showed an elevated b...  
3  The patient's weight is 65 kg. Her blood press...  
4  Weight: 70 kg\nBody temperature: 38.5°C\nBlood...




# **Data Cleaning and store in pinecone**

In [ ]:


# Create index
# pinecone.create_index(
#     name=INDEX_NAME,
#     metric='cosine',
#     dimension=1536)

/usr/local/lib/python3.10/dist-packages/pinecone/index.py:4: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm


In [ ]:

demo_index.describe_index_stats()

{'dimension': 1536,
 'index_fullness': 0.0003,
 'namespaces': {'dengue': {'vector_count': 30}},
 'total_vector_count': 30}

In [ ]:
# # Initialize our OpenAI model


In [ ]:
# # Tell tiktoken what model we'd like to use for embeddings
# tiktoken.encoding_for_model('text-embedding-ada-002')

# # Intialize a tiktoken tokenizer (i.e. a tool that identifies individual tokens (words))
# tokenizer = tiktoken.get_encoding('cl100k_base')

# # Create our custom tiktoken function
# def tiktoken_len(text: str) -> int:
#     """
#     Split up a body of text using a custom tokenizer.

#     :param text: Text we'd like to tokenize.
#     """
#     tokens = tokenizer.encode(
#         text,
#         disallowed_special=()
#     )
#     return len(tokens)

In [ ]:
def chunk_by_size(text: str, size: int = 1024):
    """
    Chunk up text recursively.

    :param text: Text to be chunked up
    :return: List of Document items (i.e. chunks).|
    """
    text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = size,
    chunk_overlap = 300,
    add_start_index = True,
)
    return text_splitter.create_documents([text])

In [ ]:
path = '/content/ent.csv'

In [ ]:
def csv_to_list(file_path):
  df_path = pd.read_csv(file_path)
  data_list = []
  for i in range(len(df_path)):
    data = f"History: \n {df_path['History'][i]}\n Primary_Daignose: \n  {df_path['Body_Vitals'][i]}\n Description \n {df_path['Description'][i]}"
    data_list.append(data)
  return data_list
chunk = csv_to_list(path)
chunk[1]

"History: \n this is a 42-year-old white female who comes in today for a complete physical and follow up on asthma. she says her asthma has been worse over the last three months. she has been using her inhaler daily. her allergies seem to be a little bit worse as well. her husband has been hauling corn and this seems to aggravate things. she has not been taking allegra daily but when she does take it, it seems to help somewhat. she has not been taking her flonase which has helped her in the past.\n Primary_Daignose: \n  {'vitals': ' her weight was 151 pounds. blood pressure is 110/60. pulse is 72. temperature is 97.1 degrees. respirations are 20.', 'general': ' this is a well-developed, well-nourished 42-year-old white female, alert and oriented in no acute distress. affect is appropriate and is pleasant.', 'heent': ' normocephalic, atraumatic. tympanic membranes are clear. conjunctivae are clear. pupils are equal, round and reactive to light. nares without turbinate edema. oropharynx 

In [ ]:
from tqdm import tqdm
from uuid import uuid4
def create_chunks_metadata_embeddings(dataset) -> list[dict]:
    """
    Given a dataset, split text data into chunks, extract metadata, create embeddings for each chunk.

    :param dataset: Data we want to process.
    :return: List of data objects to upsert into our Pinecone index.
    """
    data_objs = []

    # For each row in our dataset:
    for data in tqdm(dataset):  # (tqdm library prints status of for-loop to console)
        # Create chunks
        chunked_text = chunk_by_size(data)

        # Extract just the string content from the chunk
        chunked_text = [c.page_content for c in chunked_text]

        # Extract some metadata, create an ID, and generate an embedding for the chunk.
        # Wrap that all in a dictionary, and append that dictionary to a list (`data_objs`).
        for idx, text in enumerate(chunked_text):
            payload = {
                "metadata": {
                    "diseases": 'dengue-fever',
                    "text_content": text  # there are 248 chars in this chunk of text
                },
             "id": str(uuid4()),
            "values": embed.embed_documents([text])[0] # --> list of len 248, each item of those 248 has a len of 1536
            }
            data_objs.append(payload)

    # Return list of dictionaries, each containing our metadata, ID, and embedding, per chunk.
    return data_objs

In [ ]:
data = create_chunks_metadata_embeddings(chunk)

100%|██████████| 20/20 [00:03<00:00,  5.52it/s]


In [ ]:
data[0]

In [ ]:
BATCH_SIZE = 4

def batch_upsert(data: list[dict], index, namespace: str):
    """
    Upsert data objects to a Pinecone index in batches.

    :param data: Data objects we want to upsert.
    :param index: Index into which we want to upsert our data objects.
    :namespace: Namespace within our index into which we want to upsert our data objects.
    """
    for i in range(0, len(data), BATCH_SIZE):
        batch = data[i:i+BATCH_SIZE]
        # print(batch)
        index.upsert(vectors=batch, namespace=namespace)


In [ ]:
batch_upsert(data, demo_index, 'ent')

In [ ]:
# embed = chunk

# **JSON to CSV**

In [ ]:
import json
vas = pd.read_json("/content/pulmonologist.json").read()
# vas = vas.loads()

In [ ]:
# df.head()

In [ ]:
# df.info()

In [ ]:
# u_df = {}
# u_df['description'] = df['description']
# u_df['medical_history'] = df['medical history']
# u_df['primary_diagnose'] = df['primary diagnosis at health care center']

In [ ]:
# df_1 = pd.DataFrame(u_df)

In [ ]:
# df_1.to_csv("/content/gdrive/MyDrive/ent.csv")

In [ ]:
path = "/content/dengue_df.csv"
#"/content/gdrive/MyDrive/ent.csv"

In [ ]:
df_1 = pd.read_csv(path)

In [ ]:
df_1.head()

Unnamed: 0                                        Description  \
0           0  The patient is presenting symptoms of dengue f...   
1           1  The patient appears weak and fatigued. He has ...   
2           2  The patient is a 28-year-old female who has re...   
3           3  Upon physical examination, the patient appears...   
4           4  The patient is a 35-year-old male who presente...   

                                             History  \
0  The patient has a history of recent travel to ...   
1  The patient, a 35-year-old male, visited the c...   
2  The patient with Dengue Fever presented with a...   
3  The patient is a 38-year-old female who presen...   
4  Patient has a history of recent travel to a tr...   

                                         Body_Vitals  
0  The patient's body temperature may rise to aro...  
1  Upon examination, the patient's body temperatu...  
2  The patient's body vitals showed an elevated b...  
3  The patient's weight is 65 kg. Her blood press...  
4  Weight: 70 kg\nBody temperature: 38.5°C\nBlood...

# **Data Query**

**Medical Assessment Report Generation**

In [ ]:
# def assessment_report(messages):
#     response = client.chat.completions.create(
#       model="gpt-3.5-turbo",
#       messages=messages,
#       temperature=0.2,
#       max_tokens=1024,
#       top_p=1,
#       frequency_penalty=0.16,
#       presence_penalty=0.15
#     )
#     return response.choices[0].message.content


### **Fetching Data from Pinecone**

In [ ]:
real_query = """
History:
The person recently visited a warm place, maybe tropical or subtropical, \
where dengue fever is common. It's possible they got the dengue virus from mosquito \
bites during their trip.\
 Dengue fever is a sickness caused by a virus spread by Aedes mosquitoes.
Body Vitals:
 In the early stage of dengue fever, the person's body temperature \
 might go up to about 104°F (40°C). They might also have a strong headache,\
 along with pain in the joints and muscles, a rash, and discomfort behind the eyes.\
 As the fever continues, \
 their blood platelet count might go down, making them more likely to bleed.
 Description
 The person is showing signs of dengue fever, like a high fever, intense headache, \
 and body pain. They might have a rash on their body and mention pain behind the eyes. \
 Doctors are keeping a close eye on the patient's blood platelet count to catch any issues early. \
 The patient is getting enough fluids and rest to help them get better.
"""
def get_context(query, name_space):
  vec_query = embed.embed_documents([query])[0]
  description = demo_index.query(vector=vec_query,
                                 top_k=3, include_metadata=True,
                                 namespace=name_space)

  return (description['matches'][0]['metadata']['text_content'],
          description['matches'][1]['metadata']['text_content'],
          description['matches'][2]['metadata']['text_content'])

In [ ]:
def prompt_maker(descript1, descript2, descript3, query):
  """
  Build prompt to create first report for doctor.

  Paramters
  ---------
  descript1
    First matching of previous data.

  descript2
    First matching of previous data.
  Query
    Patient describe what he/ she feels
  """
  previous_patient = f'<<CONTEXT>> \n {descript1} \n {descript2} \n {descript3} \n <<QUERY>> \n \n {query}'
  return previous_patient



In [ ]:
context1,context2,context3 =  get_context(real_query, "dengue")

In [ ]:
patient_data = prompt_maker(context1,context2,context3, real_query)

In [ ]:
# # Prompt Template for translation.
# patient_fir = [
#             {
#               "role": "system",
#               "content": "You are a best medical knowledgde curator. You will curate the given text into presentable manner to doctor.\
#                           Keep the content revolve around the Query written under 'QUERY' tag. Kindly not use the \
#                           Context data or refernece to context data in response like age or gender from CONTEXT data.\
#                           Extract only idea from context rewrite the query in four to five sentence."
#             },
#             {
#               "role": "user"

#             }
#         ]

In [ ]:
# patient_fir[1]['content'] = patient_data
# patient_fir

In [ ]:
# FIR = assessment_report(patient_fir)
# FIR

'Based on the given information, the patient is a 28-year-old female who is experiencing a headache, fever, and weakness for the past two days. She has already taken medication but it did not provide relief. This is a new experience for her as she has not encountered these symptoms before.'

**LLM Code Query Code**



In [ ]:
asthama = """You are a exceptionally professional allergy specialist doctor. \
 Provide insights, medicines and remedies against data provided by patient. \
 Data consist of History, body vitals and description of allergy. Data in CONTEXT \
 should be use as a CONTEXT in a  final answer. \
 Answer must be in JSON format that consist of doctor's views,\
 medicine and remedies as a keys.

Here is a Data:
{input}"""


dengue_fever = """You are a exceptionally professional dengue fever specialist doctor. \
 Provide insights, medicines and remedies against data provided by patient. \
 Data consist of History, body vitals and description of allergy.Data in CONTEXT \
 should be use as a CONTEXT in a  final answer. \
 Answer must be in JSON format that consist of doctor's views,\
 medicine and remedies as a keys.

Here is a Data:
{input}"""


In [ ]:
prompt_infos = [
    {
        "name": "asthama",
        "description": "Good for answering asthema related queries",
        "prompt_template": asthama
    },
    {
        "name": "dengue_fever",
        "description": "Good for answering dengue fever related question",
        "prompt_template": dengue_fever
    }
  ]

In [ ]:
llm = ChatOpenAI(
      model="gpt-4",
      temperature=0.2,
      max_tokens=1024
    )

In [ ]:
default_prompt = ChatPromptTemplate.from_template("{input}")
default_chain = LLMChain(llm=llm, prompt=default_prompt)

In [ ]:
destination_chains = {}
for p_info in prompt_infos:
    name = p_info["name"]
    prompt_template = p_info["prompt_template"]
    prompt = ChatPromptTemplate.from_template(template=prompt_template)
    chain = LLMChain(llm=llm, prompt=prompt)
    destination_chains[name] = chain

destinations = [f"{p['name']}: {p['description']}" for p in prompt_infos]
destinations_str = "\n".join(destinations)

In [ ]:
destinations_str

'asthama: Good for answering asthema related queries\ndengue_fever: Good for answering dengue fever related question'

In [ ]:
MULTI_PROMPT_ROUTER_TEMPLATE = """When presented with a raw text input for a \
language model in the context of medical prompts,\
 identify the most appropriate model prompt for the input. \
You will receive the names of the available prompts along with descriptions indicating their\
suitability for specific purposes within the medical domain. Additionally,\
you have the option to modify the original input if you believe that revising it will result\
in a more effective response from the language model.

<< FORMATTING >>
Return a markdown code snippet with a JSON object formatted to look like:
```json
{{{{
    "destination": string \ name of the prompt to use or "DEFAULT"
    "next_inputs": string \ a potentially modified version of the original input
}}}}
```

REMEMBER: "destination" MUST be one of the candidate prompt \
names specified below OR it can be "DEFAULT" if the input is not\
well suited for any of the candidate prompts.
REMEMBER: "next_inputs" can just be the original input \
if you don't think any modifications are needed.

<< CANDIDATE PROMPTS >>
{destinations}

<< INPUT >>
{{input}}

<< OUTPUT (remember to include the ```json)>>"""

In [ ]:
router_template = MULTI_PROMPT_ROUTER_TEMPLATE.format(
    destinations=destinations_str
)
router_prompt = PromptTemplate(
    template=router_template,
    input_variables=["input"],
    output_parser=RouterOutputParser(),
)

router_chain = LLMRouterChain.from_llm(llm, router_prompt)
memory = ConversationBufferWindowMemory(llm=llm, input_key='input', output_key= 'text')
chain = MultiPromptChain(router_chain=router_chain,
                         destination_chains=destination_chains,
                         default_chain=default_chain,
                         verbose= True,
                         memory= memory,
                        )

In [ ]:
# query = """
# History:
#  23-year-old white female presents with complaint of allergies. she used to have allergies when she lived in seattle but she thinks they are worse here. in the past, she has tried claritin, and zyrtec. both worked for short time but then seemed to lose effectiveness. she has used allegra also. she used that last summer and she began using it again two weeks ago. it does not appear to be working very well. she has used over-the-counter sprays but no prescription nasal sprays. she does have asthma but doest not require daily medication for this and does not think it is flaring up.
#  Primary_Daignose:
#   {'vitals': ' weight was 130 pounds and blood pressure 124/78.', 'heent': ' her throat was mildly erythematous without exudate. nasal mucosa was erythematous and swollen. only clear drainage was seen. tms were clear.', 'neck': ' supple without adenopathy.', 'lungs': ' clear.'}
#  Description
#  a 23-year-old female presents with complaint of allergies.
#  """

# query = """
# History:
# The patient has a history of recent travel to a tropical or subtropical region
# where dengue fever is endemic. They may have been exposed to the dengue virus
#  through mosquito bites during their travel. Dengue fever is a viral infection
#  transmitted by the Aedes mosquito.

# Body Vitals:
# The patient's body temperature may rise to around 104°F (40°C) during the initial phase of
# dengue fever. Additionally, they may experience intense headache, joint and muscle pain, rash,
# and pain behind the eyes. As the fever progresses, their blood platelet count may decrease,
#  leading to an increased risk of bleeding.
# Description:
# The patient is presenting symptoms of dengue fever, such as high fever, severe headache,
# and body pain. They may have a rash on their body and complain of pain behind the eyes.
# The patient's blood platelet count is being monitored regularly to assess their condition
# and ensure early detection of any complications. Adequate hydration and rest are being
# provided to support the patient's recovery.
# """

In [ ]:
# patient_data

In [ ]:
data = chain.run(patient_data)
data.split('\n')



> Entering new MultiPromptChain chain...


/usr/local/lib/python3.10/dist-packages/langchain/chains/llm.py:316: UserWarning: The predict_and_parse method is deprecated, instead pass an output parser directly to LLMChain.
  warnings.warn(


dengue_fever: {'input': "History: The person recently visited a warm place, maybe tropical or subtropical, where dengue fever is common. It's possible they got the dengue virus from mosquito bites during their trip. Dengue fever is a sickness caused by a virus spread by Aedes mosquitoes. Body Vitals: In the early stage of dengue fever, the person's body temperature  might go up to about 104°F (40°C). They might also have a strong headache,along with pain in the joints and muscles, a rash, and discomfort behind the eyes. As the fever continues,  their blood platelet count might go down, making them more likely to bleed. Description: The person is showing signs of dengue fever, like a high fever, intense headache,  and body pain. They might have a rash on their body and mention pain behind the eyes.  Doctors are keeping a close eye on the patient's blood platelet count to catch any issues early.  The patient is getting enough fluids and rest to help them get better."}
> Finished chain.


In [ ]:
data.split('\n')

['{',
 '  "Doctor\'s Views": "Based on the patient\'s history, body vitals, and symptoms, it is highly likely that the patient is suffering from dengue fever. The recent travel to a tropical or subtropical region, the high fever, severe headache, joint and muscle pain, rash, and discomfort behind the eyes are all indicative of dengue fever. The decrease in platelet count is also a common symptom of dengue fever. It is crucial to monitor the patient\'s condition closely, especially the platelet count, to prevent complications such as dengue hemorrhagic fever and dengue shock syndrome.",',
 '',
 '  "Medicines": "There is no specific medicine to treat dengue infection. However, the patient can take pain relievers with acetaminophen and avoid medicines with aspirin, which could worsen bleeding. It is also important to stay hydrated. If the symptoms worsen, the patient should be hospitalized and might need intravenous (IV) fluids and electrolyte replacement, blood pressure monitoring, and t

In [ ]:
def doctor_prescription(textbox_value):
    data = chain.run(patient_data)
    return (data.split('\n'))

iface = gr.Interface(
    fn=doctor_prescription,
    inputs=gr.TextArea( value ="Enter Text Here", label="Patient's FIR"),
    outputs=gr.TextArea(label= "Prescription"),
    live=True,
    examples=[
        [patient_data],
    ]
)
iface.launch()


Setting queue=True in a Colab notebook requires sharing enabled. Setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://efc7aca8bb4474ec90.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)
